db_go18_test.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2017 Square Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build go1.8
+
+package squalor
+
+import (
+	"context"
+	"testing"
+	"time"
+)
+
+func TestWithContextTimeout(t *testing.T) {
+	db := makeTestDB(t, usersDDL)
+	defer db.Close()
+
+	if db.GetContext() == nil {
+		t.Fatal("Expected default non-nil Context on DB")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	defer cancel()
+
+	db = db.WithContext(ctx).(*DB)
+
+	startTime := time.Now()
+
+	// This query should be canceled.
+	if _, err := db.Query("SELECT SLEEP(1)"); err != context.DeadlineExceeded {
+		t.Fatalf("expected context.DeadlineExceeded, got %v", err)
+	}
+	if d := time.Since(startTime); d > 500*time.Millisecond {
+		t.Errorf("too long execution time: %s", d)
+	}
+
+	// This query should immediately err because the context has already exceeded
+	// the timeout.
+	if _, err := db.Query("SELECT 1"); err != context.DeadlineExceeded {
+		t.Fatalf("expected context.Canceled, got %v", err)
+	}
+}
+
+func TestWithContextDeadline(t *testing.T) {
+	db := makeTestDB(t, usersDDL)
+	defer db.Close()
+
+	if db.GetContext() == nil {
+		t.Fatal("Expected default non-nil Context on DB")
+	}
+
+	startTime := time.Now()
+	d := startTime.Add(100 * time.Millisecond)
+
+	ctx, cancel := context.WithDeadline(context.Background(), d)
+	defer cancel()
+
+	db = db.WithContext(ctx).(*DB)
+
+	// This query should be canceled.
+	if _, err := db.Query("SELECT SLEEP(1)"); err != context.DeadlineExceeded {
+		t.Fatalf("expected context.DeadlineExceeded, got %v", err)
+	}
+	if d := time.Since(startTime); d > 500*time.Millisecond {
+		t.Errorf("too long execution time: %s", d)
+	}
+
+	// This query should immediately err because the context has already exceeded
+	// the deadline.
+	if _, err := db.Query("SELECT 1"); err != context.DeadlineExceeded {
+		t.Fatalf("expected context.Canceled, got %v", err)
+	}
+}
+
+func TestWithContextNoTimeout(t *testing.T) {
+	db := makeTestDB(t, usersDDL)
+	defer db.Close()
+
+	if db.GetContext() == nil {
+		t.Fatal("Expected default non-nil Context on DB")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
```

Comment:
My other comment about timing during unit tests applies here too. While I would expect a select statement to execute almost instantly, it might not. Or it might not run until after the timeout has already expired.

To reduce this chance of this being a problem, I'd set the timeout to 1 minute.